add support for avrtiny devices

pdenes opened the pull request:
I was looking for something exactly like this library for an attiny10. It's perfect! Only had to make a few small changes to make it work with a "reduced tiny" core:
- use `__tmp_reg__` (alias for r0, but r16 on avrtiny)
- use r25 instead of r18 on avrtiny
- avrtiny has no `lpm`, but can simply be replaced by `ld`

pdenes committed:
add support for avrtiny devices

nerdralph commented:
The __tmp_reg__ changes are good, and in retrospect I probably should've used it in the beginning.
I prefer to avoid conditional compilation code where possible.   I see r18 is a call-saved register for avrtiny, so it shouldn't be used in the asm code for parameter passing.  Why did you pick r25?  I'm thinking I could change r18 to r20, which is call-clobbered on both.

pdenes commented:
That sounds better indeed! My only reasoning was to keep everything unchanged for the non-avrtiny case, to minimise the impact, but it has the price of those conditionals. r25 just happens to be one that's free (r20 and r21 are already used).
There is also the issue of renaming labels, etc. that reference registers. I'm happy to play with this more, following what you suggest or try out anything, but at the moment can only test on an attiny10.

nerdralph commented:
I don't have any tiny core AVRs to test with on my side.  Having a single code path should simplify testing.  If it builds OK for the tiny core, and the regular AVR build works a couple AVR MCUs that I have (i.e. t13, m328), then it will probably work on the tiny core AVR too.

pdenes committed:
change registers so that the same set is used

- move registers: r18 -> r20, r20,r21 -> r22,r23
- rename functions, etc. accordingly
- remove #define TX_REG

pdenes commented:
Here's an update: I changed all registers to a common set and got rid of that #define. Hopefully got it right - works fine on the attiny10. The only conditional that remains is for accessing progmem, but that seems unavoidable.

pdenes committed:
fix one comment

nerdralph commented:
I thought there was a macro in AVR-libc or the gnu assembler for ld/lpm that would expand to the correct opcode for the target.  After looking through the source, I couldn't find anything.  If there is one, that would avoid the "#if defined(__AVR_TINY__)" code.
The comment about using r18 should be changed to r20:
https://github.com/pdenes/debugSerial/blob/avrtiny/print.S#L3

Otherwise the code looks OK.  I plan to clone your branch and test it on a t85 and t13a in the next few days before I merge it.
